### ISSUE TITLE
Default Kafka Channel Config does not create Kafka Topic

### ISSUE THREAD

**Issue by kameshsampath** [knative/eventing-contrib#1239]
**Describe the bug**
After configuring default-ch-webhook as described https://knative.dev/docs/eventing/channels/default-channels/, creating a `Channel` on custom namespace does not create Kafka Channel and hence no Kafka Topic gets created.

 **Expected behavior**
A KafkaChannel and a corresponding KafkaTopic is expected to be create Kafka channel called `my-events-ch` and a KafkaTopic named `knative-messaging-kafka.knativetutorial.my-events-ch`

 **To Reproduce**

**Assuming Cluster as Knative CRDS + Knative Serving/Eventing Components installed.**

1. ```shell
kubectl apply \
-f https://github.com/knative/eventing-contrib/\
releases/download/v0.14.1/kafka-source.yaml
```

2. ```shell
curl -L "https://github.com/knative/eventing-contrib/\
releases/download/v0.14.1/kafka-channel.yaml" \
 | sed 's/REPLACE_WITH_CLUSTER_URL/my-cluster-kafka-bootstrap.kafka:9092/' \
 | kubectl apply --filename -
```

3. ```shell
cat <<EOF | kubectl create -f - 
apiVersion: v1
kind: ConfigMap
metadata:
  name: default-ch-webhook
  namespace: knative-eventing
data:
  default-ch-config: |
    clusterDefault: 
      apiVersion: messaging.knative.dev/v1alpha1
      kind: InMemoryChannel
    namespaceDefaults: 
      mynamespace:
        apiVersion: messaging.knative.dev/v1alpha1
        kind: KafkaChannel
        spec:
          numPartitions: 1
          replicationFactor: 1
EOF```

 **Knative release version**
v0.14. of Serving and Eventing
v0.14.1 of Eventing Contrib

**Comment by kameshsampath** [comment]
@matzew @slinkydeveloper  ^^

**Comment by slinkydeveloper** [comment]
I'm unable to reproduce the error:

1. I've installed kn with https://github.com/matzew/kn-box
2. Created the namespace with `kubectl create namespace knativetutorial`
3. Applied this CM:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: default-ch-webhook
  namespace: knative-eventing
data:
  default-ch-config: |
    clusterDefault:
      apiVersion: messaging.knative.dev/v1alpha1
      kind: InMemoryChannel
    namespaceDefaults:
      knativetutorial:
        apiVersion: messaging.knative.dev/v1alpha1
        kind: KafkaChannel
        spec:
          numPartitions: 1
          replicationFactor: 1
```

4. Created the channel:

```shell
% kubectl apply -f - << EOF
apiVersion: messaging.knative.dev/v1beta1
kind: Channel
metadata:
  name: "test"
  namespace: knativetutorial
spec: {}
EOF
channel.messaging.knative.dev/test created
```

5. Channel ready:

```shell
% kubectl get kafkachannels.messaging.knative.dev -n knativetutorial
NAME   READY   REASON   URL                                                        AGE
test   True             http://test-kn-channel.knativetutorial.svc.cluster.local   44s
% kubectl get kt --all-namespaces
NAMESPACE   NAME                                           PARTITIONS   REPLICATION FACTOR
kafka       knative-messaging-kafka.knativetutorial.test   1            1
```

**Comment by kameshsampath** [comment]
Can you try with a Kind Cluster? That’s what I was using today

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by slinkydeveloper** [comment]
I never managed to make kind working on my machine, maybe @lionelvillard has more inputs on that

**Comment by pierDipi** [comment]
/assign

I cannot reproduce the issue.

Steps I followed:

1. create a kind cluster
```bash
kind create cluster --name kafkabug
kubectl cluster-info --context kind-kafkabug

```

2. set up kafka using strimzi operator: `kafka_setup`
```bash
#!/usr/bin/env bash
set -e
# Turn colors in this script off by setting the NO_COLOR variable in your
# environment to any value:
#
# $ NO_COLOR=1 test.sh
NO_COLOR=${NO_COLOR:-""}
if [ -z "$NO_COLOR" ]; then
  header=$'\e[1;33m'
  reset=$'\e[0m'
else
  header=''
  reset=''
fi
strimzi_version=`curl https://github.com/strimzi/strimzi-kafka-operator/releases/latest |  awk -F 'tag/' '{print $2}' | awk -F '"' '{print $1}' 2>/dev/null`
function header_text {
  echo "$header$*$reset"
}
header_text "Using Strimzi Version:                  ${strimzi_version}"
header_text "Strimzi install"
kubectl create namespace kafka
curl -L "https://github.com/strimzi/strimzi-kafka-operator/releases/download/${strimzi_version}/strimzi-cluster-operator-${strimzi_version}.yaml" \
  | sed 's/namespace: .*/namespace: kafka/' \
  | kubectl -n kafka apply -f -
header_text "Applying Strimzi Cluster file"
kubectl -n kafka apply -f "https://raw.githubusercontent.com/strimzi/strimzi-kafka-operator/${strimzi_version}/examples/kafka/kafka-ephemeral-single.yaml"
```
3. apply eventing and Kafka channel
```bash
#!/usr/bin/env bash

kubectl apply --selector knative.dev/crd-install=true --filename https://github.com/knative/eventing/releases/download/v0.14.0/eventing.yaml

kubectl apply --filename https://github.com/knative/eventing/releases/download/v0.14.0/eventing.yaml

# run kafka with strimzi
./kafka_setup

curl -L "https://github.com/knative/eventing-contrib/\
releases/download/v0.14.1/kafka-channel.yaml" \
 | sed 's/REPLACE_WITH_CLUSTER_URL/my-cluster-kafka-bootstrap.kafka:9092/' \
 | kubectl apply --filename -

kubectl create ns mynamespace
```
4. apply the following yaml
```yaml
apiVersion: v1
data:
  default-ch-config: |
    clusterDefault:
      apiVersion: messaging.knative.dev/v1alpha1
      kind: InMemoryChannel
    namespaceDefaults:
      mynamespace:
        apiVersion: messaging.knative.dev/v1alpha1
        kind: KafkaChannel
        spec:
            numPartitions: 1
            replicationFactor: 1
kind: ConfigMap
metadata:
  annotations:
  labels:
    eventing.knative.dev/release: devel
  name: default-ch-webhook
  namespace: knative-eventing
---
apiVersion: messaging.knative.dev/v1beta1
kind: Channel
metadata:
  name: "test"
  namespace: mynamespace
spec: {}
```
5. check Kafka channels
```bash
kubectl get kafkachannels.messaging.knative.dev -n mynamespace 
```

Output
```bash
NAMESPACE     NAME   READY   REASON   URL                                                    AGE
mynamespace   test   True             http://test-kn-channel.mynamespace.svc.cluster.local   25s
```

**Comment by kameshsampath** [comment]
@pierDipi  - thanks for trying out, let me try again and let you know. If thats not occuring I will close the issue

**Comment by kameshsampath** [comment]
@slinkydeveloper  @pierDipi  - it works now. The core issue was that when I created I did create the channel using `messaging.knative.dev/v1alpha1` which was not having any reconcilation to be fired, but when I changed the channel def to using `messaging.knative.dev/v1beta1` it picked up the channel for reconciliation and working as expected. 

```yaml
---
apiVersion: messaging.knative.dev/v1beta1
kind: Channel
metadata:
  name: my-events-ch
  namespace: knativetutorial
spec: {}
```

Wondering we should callout using right *apiVersion* of the API in the docs ?

**Comment by slinkydeveloper** [comment]
I think so, mind doing a PR @kameshsampath ?

**Comment by kameshsampath** [comment]
You mean for Knative Docs @slinkydeveloper  ? I should be able to

**Comment by slinkydeveloper** [comment]
@kameshsampath can we close this one?

**Comment by kameshsampath** [comment]
Yep we could.  Thanks.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

________________________________
